### ISSUE TITLE
Diamond shaped graphs

### ISSUE THREAD

**Issue by farrutapps** [facebookexperimental/libunifex#220]
Hello, I have been following your great work and would like to try it in a project. I now wanted to create a diamond-shaped asynchronous graph but did not see how to do this with the available algorithms.

Is there any such? If not, how could this be done conceptually using sender/receiver?

In particular, the graph looks like follows:

           A
           |
           B
        /     \
       C       D
       \       /
           E 

The nodes `A` through `E` represent senders and the edges are some computations connecting two senders. Assume `B-C` and `B-D` shall be computed asynchronously and are different operations, both using the result from `B`. I reckon `E` would be computed as `E = when_all(C,D)`.

How could the "splitting" at `B` be done?

**Comment by ericniebler** [comment]
`when_all` is precisely such a fork/join.  Does it not meet your needs?

**Comment by farrutapps** [comment]
I'm afraid I don't quite see the "fork" part of `when_all`. Given the sender `B` how can I use `when_all` to produce two new senders, `C`, `D` which reuse the value sent by `B`?

The naive example below, which just uses the sender variable `B` twice, results in two invocations of `foo`. However, I would like `foo` to be executed only once.
```cpp
sender auto A = // get from somewhere
auto B = A | unifex::transform(foo);
auto C = B | unifex::transform(bar);
auto D = B | unifex::transform(other);
auto E = when_all(C, D);
```

**Comment by apmccartney** [comment]
Hi @farrutapps

> I'm afraid I don't quite see the "fork" part of when_all

I have a similar understanding. This is the way I model these situations:

```
sender auto A = // get from somewhere
sender auto B = A | unifex::transform(foo);
sender auto F = B | unifex::let_value([&](const auto&... args) { 
   /** 
    * Pessimistically capturing by reference. This is safe because `let_value` provides lifetime extension (whether you need it or not).
    *
    * You may be able to capture by value in your use case.
    * /
    sender auto C = unifex::just(std::ref(args)...) |  unifex::transform(bar); 
    sender auto D = unifex::just(std::ref(args)...) | unifex::transform(other);
    auto E = unifex::when_all(std::move(C), std::move(D));
});
```

**Comment by farrutapps** [comment]
Hi @apmccartney, thanks a lot for your input, this is pretty much what I have been looking for. Thanks!

**Comment by fililili** [comment]
> 
hi, I find that [P2300r1](https://wg21.link/P2300r1) has mentioned split function, but split is not implemented yet, is split function suitable for such situation?

**Comment by ccotter** [comment]
I modified P2300's reference implementation of `split` to work with libunifex, with the caveat that my modified version allows Senders of move-only types to be split and connected to multiple times. I've got a rough draft [pushed to my fork](https://github.com/facebookexperimental/libunifex/compare/main...ccotter:libunifex:split). I've been using this version successfully to create a "shared_task" aka `decltype(split(task<T>))`. It's not ready for upstreaming immediately, but could be done if the libunifex maintainers are interested.
